File: lab03/py/src/models/book.py
import os
from abc import ABC, abstractmethod
from dataclasses import dataclass
from pathlib import Path

type BookFiles = dict[str, str]


@dataclass
class Chapter:
    title: str
    url: str
    file_index: int
    start_time: int
    end_time: int

    @property
    def duration(self) -> int:
        return self.end_time - self.start_time


@dataclass(kw_only=True)
class BookSource(ABC):
    url: str = ""
    cover: str = ""

    @property
    @abstractmethod
    def dir_path(self) -> Path:
        pass

    @property
    def cover_path(self) -> Path:
        return Path("cover.jpg")


@dataclass(kw_only=True)
class TextBook(BookSource):
    publication: str
    file_url: str

    @property
    def dir_path(self) -> Path:
        return Path(".")


@dataclass(kw_only=True)
class AudioBook(BookSource):
    narrator: str
    chapters: list[Chapter]

    @property
    def dir_path(self) -> Path:
        return Path(".", self.narrator)


@dataclass(kw_only=True)
class BookData:
    title: str
    author: str
    series_name: str
    number_in_series: str

    @property
    def book_path(self) -> Path:
        """
        :returns: Relative path to the book in the library
        """
        path = Path(".", self.author)
        if self.series_name:
            book_name = (
                f"{str(self.number_in_series).rjust(2, '0')}. {self.title}"
            )
            path /= self.series_name
            path /= book_name
        else:
            path /= self.title
        return path

    @property
    def dir_path(self) -> Path:
        """
        :returns: Absolute path to the directory where the book is stored
        """
        return Path(os.environ["books_folder"], self.book_path).absolute()

    @property
    def cover_path(self) -> Path:
        """
        :returns: Absolute path to the book cover file
        """
        return self.dir_path / "cover.jpg"


@dataclass(kw_only=True)
class RawBook[SourceT: BookSource](BookData):
    source: SourceT

    @property
    def dir_path(self) -> Path:
        return super().dir_path / self.source.dir_path

    def __format__(self, format_spec: str, /) -> str:
        return repr(self)

File: lab03/cs/without/Models/Book.cs
namespace BookDownloader.Models;

public class Chapter
{
    public string Title { get; set; }
    public string Url { get; set; }
    public int FileIndex { get; set; }
    public int StartTime { get; set; }
    public int EndTime { get; set; }

    public int Duration => EndTime - StartTime;
}

public abstract class BookSource
{
    public string Url { get; set; } = "";
    public string Cover { get; set; } = "";

    public abstract string DirPath { get; }
    public virtual string CoverPath => "cover.jpg";
}

public class TextBook : BookSource
{
    public string Publication { get; set; }
    public string FileUrl { get; set; }

    public override string DirPath => ".";
}

public class AudioBook : BookSource
{
    public string Narrator { get; set; }
    public List<Chapter> Chapters { get; set; } = new List<Chapter>();

    public override string DirPath => Path.Combine(".", Narrator);
}

public class BookData
{
    public string Title { get; set; }
    public string Author { get; set; }
    public string SeriesName { get; set; }
    public string NumberInSeries { get; set; }

    public string BookPath
    {
        get
        {
            var path = Path.Combine(".", Author);
            if (!string.IsNullOrEmpty(SeriesName))
            {
                var bookName = $"{NumberInSeries.PadLeft(2, '0')}. {Title}";
                path = Path.Combine(path, SeriesName, bookName);
            }
            else
            {
                path = Path.Combine(path, Title);
            }
            return path;
        }
    }

    public string DirPath
    {
        get
        {
            var booksFolder = Environment.GetEnvironmentVariable("books_folder");
            return Path.GetFullPath(Path.Combine(booksFolder, BookPath));
        }
    }

    public string CoverPath => Path.Combine(DirPath, "cover.jpg");
}

public class RawBook : BookData
{
    public BookSource Source { get; set; }

    public string DirPath => Path.Combine(base.DirPath, Source.DirPath);
}
